### PR TITLE
Make never consistent with always

### PR DESCRIPTION
**Pull request by Luis-Henriquez-Perez**
As it stands, when you adding multiple conditions in the `always` command making the condition harder to return t (in general, that is obviously it depends on the conditions added).

```elisp
(loopy (list i '(1 0 1 0 1))
       (always (< i 2) (>= i 0)))

;; The set of numbers that meet the condition (and (< i 2) (>= i 0)) is less
;; than that of the set of numbers that meet the condition (< i 2).
```

However, for never, adding more conditions makes the set larger.

```elisp
(loopy (list i '(1 0 1 0 1))
       (never (< i 2) (>= i 0)))
;; The set of numbers that meet the condition (not (and (< i 2) (>= i 0))) is
;; greater than that of just (not (< i 2))
```

I think that that's why we should use `or` for `never` instead of the implied `and` so that it will also (in general) lessen the set of candidates. What do you thinK?

Also in this PR I elaborate on example of `always` (and I plan to also do so for `never).

### CODE REVIEW

**Comment by okamsn** [comment]
> As it stands, when you adding multiple conditions in the `always` command making
> the condition harder to return t (in general, that is obviously it depends on
> the conditions added).
>
> ```elisp
> (loopy (list i '(1 0 1 0 1))
>        (always (< i 2) (>= i 0)))
>
> ;; The set of numbers that meet the condition (and (< i 2) (>= i 0)) is less
> ;; than that of the set of numbers that meet the condition (< i 2).
> ```
>
> However, for never, adding more conditions makes the set larger.
>
> ```elisp
> (loopy (list i '(1 0 1 0 1))
>        (never (< i 2) (>= i 0)))
> ;; The set of numbers that meet the condition (not (and (< i 2) (>= i 0))) is
> ;; greater than that of just (not (< i 2))
> ```
>
> I think that that's why we should use `or` for `never` instead of the implied
> `and` so that it will also (in general) lessen the set of candidates. What do
> you thinK?

This makes sense to me, that since `never` requires that all conditions are false,
it should exit when any conditions are true.

> Also in this PR I elaborate on example of `always` (and I plan to also do so for
> `never).

OK.  Let me know when you are happy with it, and I will merge this PR.

**Comment by Luis-Henriquez-Perez** [comment]
Ok. I've happy these changes now.

One question: I noticed that we test to see if there is only one condition. However, using `(and ,@conditions)` with just one condition would evaluate to the same result `1 <=> (and 1)`. Why is this?

```elisp
,(if (= 1 (length conditions)) (cl-first conditions) `(and ,@conditions))
```

Is it to make macroexpansion is more deliberate and readable? If that is the case perhaps we should account for the case of 0 conditions as well.

**Comment by okamsn** [comment]
> Ok. I've happy these changes now.
>
> One question: I noticed that we test to see if there is only one
> condition. However, using `(and ,@conditions)` with just one condition would
> evaluate to the same result `1 <=> (and 1)`. Why is this?
>
> ```elisp
> ,(if (= 1 (length conditions)) (cl-first conditions) `(and ,@conditions))
> ```
>
> Is it to make macroexpansion is more deliberate and readable?

It's to make the expansion more efficient.

> If that is the case perhaps we should account for the case of 0 conditions as
> well.

For these particular commands, I don't think it makes sense to try to use them
without at least one argument (correct me if you think otherwise). I would guess
that trying to test 0 conditions is more likely to be a mistake.

**Comment by okamsn** [comment]
> > If that is the case perhaps we should account for the case of 0 conditions as
> > well.
> 
> For these particular commands, I don't think it makes sense to try to use them
> without at least one argument (correct me if you think otherwise). I would guess
> that trying to test 0 conditions is more likely to be a mistake.

Loop and Iterate both require a condition for their clauses, so I think it's safe for us to do so too. For now, I've made that change. Please let me know if there is a use case for having no conditions, and I will undo this change.

Thank you for these improvements.

**Comment by Luis-Henriquez-Perez** [comment]
> Loop and Iterate both require a condition for their clauses, so I think it's safe for us to do so too. For now, I've made that change. Please let me know if there is a use case for having no conditions, and I will undo this change.

I think there are merits to both ways of doing it.

One could argue that allowing no conditions is more conventional with the behaviour of many forms in lisp. For example, `and` and `or` which evaluate to t and nil respectively if no arguments are specified even though it doesn't make sense to use them without arguments.

On the other hand, allowing things that don't make sense can lead to bugs and typos or at indirect code that relies on a obscure behaviour (for example, relying on the behavior of `and` to return `t` in the case of 0 arguments instead of explicit accounting for that case in a macro).

In a nutshell,

I think that if we allow no conditions, we should document this case.

However, if we choose treat this case as a user mistake or as something improper that should not be done then I think we should go all the way and raise an error if the condition is not specified.
